### PR TITLE
goto_symex: rename type symbols only when needed

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -788,12 +788,62 @@ void goto_symex_statet::rename_address(
...
 
+static bool needs_renaming(
+  const typet &type,
+  const namespacet &ns)
+{
+  if(type.id()==ID_array)
+  {
+    const auto &array_type=to_array_type(type);
+    return needs_renaming(array_type.subtype(), ns) ||
+           !array_type.size().is_constant();
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp:640]: This case, and also the struct/union case are only almost the same as the actual renaming function. Such subtle differences look dangerous.

File: src/goto-symex/goto_symex_state.cpp [hunk 2]
```diff
@@ -788,12 +788,62 @@ void goto_symex_statet::rename_address(
...
+}
+
 void goto_symex_statet::rename(
   typet &type,
   const irep_idt &l1_identifier,
   const namespacet &ns,
   levelt level)
 {
+  if(!needs_renaming(type, ns))
+    return;
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: That will fail to do `l1_type_entry.first->second=type;`

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/goto-symex/goto_symex_state.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-symex/goto_symex_state.cpp:640]: Fixed by means of an extra commit that first does cleanup.

File: src/goto-symex/goto_symex_state.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-symex/goto_symex_state.cpp]: Fixed.

**Review by tautschnig — APPROVED** [review]

**Review by tautschnig — CHANGES_REQUESTED** [review]
Well, it looked ok - but tests are failing, notably tests involving bit fields.

File: src/goto-symex/goto_symex_state.cpp [hunk 3]
```diff
@@ -689,14 +683,14 @@ void goto_symex_statet::rename(
   }
   else if(type.id() == ID_union_tag)
   {
-    const symbolt &symbol = ns.lookup(to_union_tag_type(type));
-    type = symbol.type;
+    const auto &union_type = ns.follow_tag(to_union_tag_type(type));
+    type = union_type;
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex_state.cpp]: it seems `union_type` is not used aside from that copy, so type can be copied directly from the `follow_tag` result

File: src/goto-symex/goto_symex_state.cpp [hunk 4]
```diff
@@ -689,14 +683,14 @@ void goto_symex_statet::rename(
...
     rename(type, l1_identifier, ns, level);
   }
   else if(type.id() == ID_struct_tag)
   {
-    const symbolt &symbol = ns.lookup(to_struct_tag_type(type));
-    type=symbol.type;
+    const typet &struct_type = ns.follow_tag(to_struct_tag_type(type));
+    type = struct_type;
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex_state.cpp]: same thing as above

File: src/goto-symex/goto_symex_state.cpp [hunk 5]
```diff
@@ -660,22 +660,16 @@ void goto_symex_statet::rename(
     rename(array_type.subtype(), irep_idt(), ns, level);
     rename(array_type.size(), ns, level);
   }
-  else if(type.id()==ID_struct ||
-          type.id()==ID_union ||
-          type.id()==ID_class)
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex_state.cpp:670]: This is not refactoring as the commit message suggests and this change causes some tests (in particular jbmc/regression/jbmc-strings/StringSubstring) to fail.

File: src/goto-symex/goto_symex_state.cpp [hunk 6]
```diff
@@ -660,22 +660,16 @@ void goto_symex_statet::rename(
...
     struct_union_typet::componentst &components=s_u_type.components();
 
-    for(struct_union_typet::componentst::iterator
-        it=components.begin();
-        it!=components.end();
-        ++it)
-      // be careful, or it might get cyclic
-      if(it->type().id()==ID_array)
-        rename(to_array_type(it->type()).size(), ns, level);
-      else if(it->type().id()!=ID_pointer)
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex_state.cpp:682]: This change doesn't look like refactoring

**Comment by tautschnig** [comment]
@kroening Is this still relevant?

**Comment by TGWDB** [comment]
Closing as this has been untouched for a long time and no response to author query circa Feb. 2019. Please reopen if you believe this is erroneous.

**Pull request closed without merging**